### PR TITLE
[Instrumentation.StackExchangeRedis] Metrics support

### LINKED ISSUE #1742: Metric instrumentation for Redis client
Similar to: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs

### PR DESCRIPTION
**Pull request by tiagodaraujo**
Fixes #1742
Design discussion issue #

## Changes

Add the StackExchangeRedis Meter

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by matt-hensley** [comment]
New database metric semantic conventions are being worked on and should be stabilized soon. It doesn't look what was proposed in open-telemetry/opentelemetry-specification#2070 made it to the spec

You can see the current progress in the semantic convention repo:
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-metrics.md

It's been requested to not implement the new metrics until they are marked as stable.

That being said, the work here is useful, as converting `redis.client.request.duration` to the new `db.client.operation.duration` looks straightforward.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 1]
```diff
@@ -135,55 +145,93 @@ internal static class RedisProfilerEntryToActivityConverter
...
+                activity.AddEvent(new ActivityEvent("Completion", completion));
             }
 
             options.Enrich?.Invoke(activity, command);
         }
 
-        activity.Stop();
+        if (metrics.Enabled && meterTags is TagList meterTagList)
+        {
+            metrics.QueueHistogram.Record(command.EnqueuedToSending.TotalSeconds, meterTagList.ToArray());
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:211]: why `ToArray()` is needed?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 2]
```diff
@@ -88,30 +92,36 @@ internal static class RedisProfilerEntryToActivityConverter
...
-        if (activity == null)
+        if (activity is null && metrics.Enabled is false)
         {
             return null;
         }
 
-        activity.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        activity?.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        var meterTags = metrics.Enabled ?
+            (IList<KeyValuePair<string, object?>>)new TagList(StackExchangeRedisConnectionInstrumentation.CreationTags.ToArray()) :
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:102]: You can use TagList itself, no need of allocating array.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,15 @@
...
 ## Unreleased
 
+## 1.9.0-beta.2
+
+Released 2024-Jul-30
+
+* Add `OpenTelemetry.Instrumentation.StackExchangeRedis` Meter
+([#1982](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1982))
+* New Metrics: `redis.client.request.duration`,
+`redis.client.request.waiting_response`, `redis.client.request.time_in_queue`
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/CHANGELOG.md:12]: Please add entry under unreleased section. The exact version and the release date will be handled later (separate PR).

```suggestion
* Add `OpenTelemetry.Instrumentation.StackExchangeRedis` meter. Supported metrics:
  * redis.client.request.duration`,
  * `redis.client.request.waiting_response`,
  * `redis.client.request.time_in_queue`.
  ([#1982](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1982))
```

**Review by Kielek — COMMENTED** [review]
Could you please fix dotnet-format issues?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 1 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: Good catch, I added it to a unit test validation and forgot this ToArray(). Removed.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 2 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: True. TagList doesn't accept an `IEnumerable<KeyValuePair<string, object?>>`. Instead, I changed the `static readonly CreationTags` type to an array `KeyValuePair<string, object?>[]`.

Explicit cast removed. Thanks

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/CHANGELOG.md (hunk 3 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/CHANGELOG.md]: Done, can you check, please? Thanks

Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/CHANGELOG.md]: Formatting also can be adjusted. See my previous comment.

**Comment by tiagodaraujo** [comment]
@matt-hensley, like you said, at least the `redis.client.request.duration` follows the convention.

I changed all the attributes, including the tracing ones, paying attention to the 1.26.0 db convention.

There is nothing in the convention for the queue and server time, so I suggest `redis.client.queue.duration` and `redis.client.network.duration`. They may be useful in understanding `redis.client.request.duration` better as it can fluctuate depending on the queue time or the waiting time for the server.

A chart to try to show the request, queue and network flows.
```
client.request |---------------------------|
client.queue     |----|
client.network        |-------------------|
```

We could have the processing time on the client which would be: client.processing = client.request - client.queue - client.network. But I'll leave this one aside for discussion if you think it's relevant.

What do you think about this and what can we do next?

Thanks in advance

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/CHANGELOG.md (hunk 3 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/CHANGELOG.md]: My bad, I missed the formatting. Done.

**Comment by Kielek** [comment]
@tiagodaraujo, for future changes, could you please add additional commits to the PR instead of force upgrades? It is a bit easier to follow. While merging we are squashing everything to one commit.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+using System.Diagnostics.Metrics;
+using System.Reflection;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation;
+
+internal class RedisMetrics : IDisposable
+{
+    internal const string MetricRequestDurationName = "redis.client.request.duration";
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:12]: Is this expected to follow https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-metrics.md ?

File: src/Shared/SemanticConventions.cs [hunk 5]
```diff
@@ -55,7 +55,7 @@ internal static class SemanticConventions
     public const string AttributeDbInstance = "db.instance";
     public const string AttributeDbCassandraKeyspace = "db.cassandra.keyspace";
     public const string AttributeDbHBaseNamespace = "db.hbase.namespace";
-    public const string AttributeDbRedisDatabaseIndex = "db.redis.database_index";
+    public const string AttributeDbRedisFlagsKeyName = "db.redis.flags";
```
Comment (cijothomas) [inline src/Shared/SemanticConventions.cs:58]: Where is this attribute name coming from ?  I am not able to locate it in semantic conventions : https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/redis.md

Comment (tiagodaraujo) [reply src/Shared/SemanticConventions.cs:58]: The attribute already exists in the current tracing implementation. Although, I didn't locate it at the semantic convention, like you said. However, I kept it. Does it make sense to keep it and add to the semantic convention or remove in the implementation (both trace and metrics)?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 4 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:12]: The naming of `redis.client.request.duration` follows the convention.

I'm sorry, I'm not sure I understand the question.

File: src/Shared/SemanticConventions.cs (hunk 5 above)
Comment (cijothomas) [reply src/Shared/SemanticConventions.cs:58]: If something does not exist in semantic convention, then its better to remove it. 
(It is recommended to open an issue to track removing it first, and then do the actual removal)

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 4 above)
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:12]: the name from the semantic convention doc is "db.client.operation.duration" right? My question was, how did we pick the name "redis.client.request.duration" ?

File: src/Shared/SemanticConventions.cs (hunk 5 above)
Comment (tiagodaraujo) [reply src/Shared/SemanticConventions.cs:58]: issue created: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/2000

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 4 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: got it. sorry for my confusion, there are many implementations where the convention is not followed and this confused me, for example, cassandra (_db=cassandra_.client._operation=request_.duration).

I changed it to db.client.operation.duration.

Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: @cijothomas, can you check, please?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: Can you confirm if this is expected to follow https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-metrics.md, so I can review accordingly?

Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: Yes, this follows as expected. `db.client.operation.duration`. Could you review it, please?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+using System.Reflection;
+using OpenTelemetry.Internal;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation;
+
+internal class RedisMetrics : IDisposable
+{
+    internal const string OperationMetricName = "db.client.operation.duration";
+    internal const string QueueMetricName = "db.client.queue.duration";
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:14]: Are these extra metrics, not yet part of https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-metrics.md?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 7]
```diff
@@ -88,102 +92,148 @@ static bool GetCommandAndKey(
...
+        if (activity is not null)
+        {
             if (options.SetVerboseDatabaseStatements)
             {
                 var (commandAndKey, script) = MessageDataGetter.Value.Invoke(command);
 
                 if (!string.IsNullOrEmpty(commandAndKey) && !string.IsNullOrEmpty(script))
                 {
-                    activity.SetTag(SemanticConventions.AttributeDbStatement, commandAndKey + " " + script);
+                    activity.SetTag(SemanticConventions.AttributeDbQueryText, commandAndKey + " " + script);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:134]: is this intended to be part of this PR? If this is unrelated to adding Metrics, please make it separate PR, so it is easy to review changes.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 8]
```diff
@@ -88,102 +92,148 @@ static bool GetCommandAndKey(
...
+                var ip = ipEndPoint.Address.ToString();
+                var port = ipEndPoint.Port;
+
+                activity?.SetTag(SemanticConventions.AttributeServerAddress, ip);
+                activity?.SetTag(SemanticConventions.AttributeServerPort, port);
+                activity?.SetTag(SemanticConventions.AttributeNetworkPeerAddress, ip);
+                activity?.SetTag(SemanticConventions.AttributeNetworkPeerPort, port);
+
+                meterTags?.Add(SemanticConventions.AttributeServerAddress, ip);
+                meterTags?.Add(SemanticConventions.AttributeServerPort, port);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:166]: ServerPort is only to be populated if it is not the default one, right?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 9]
```diff
@@ -88,102 +92,148 @@ static bool GetCommandAndKey(
...
-            // command.SentToResponse;          // 00:00:00.0060586
-            // command.ResponseToCompletion;    // 00:00:00.0002601
+        // Total:
+        // command.ElapsedTime;             // 00:00:32.4988020
 
-            // Total:
-            // command.ElapsedTime;             // 00:00:32.4988020
+        var flags = command.Flags.ToString();
+        activity?.SetTag(SemanticConventions.AttributeDbRedisFlagsKeyName, flags);
+        meterTags?.Add(SemanticConventions.AttributeDbRedisFlagsKeyName, flags);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:120]: this is not part of metric tags, so should not be added.

**Comment by cijothomas** [comment]
@tiagodaraujo Thanks for working on this. Can you ensure this PR focuses on one aspect only? If the goal is add metrics support, then leave out all other changes to Activity, as that is making it hard to review. Its totally okay (and encouraged) to send those as separate, smaller PRs, so its easier to focus on this PR with just metrics only

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 6 above)
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: Correct. Its best to start adding those metrics part of the conventions. Once sem. conventions add more, we can come back and add more here too.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 7 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: understood, I thought I would give my maximum contribution to this PR but I started to understand the complexity of managing naming, conventions and following guidelines. Without it, it could be a mess.

I removed all changes that applied to activities, in the next PR I will explore the conventions for tracing and their implementation. Sorry for any confusion

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 9 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: removed

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 8 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:162]: is there any reason not to populate it when it is the default port?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:162]: That is what the semantic conventions states.
Also, every additional attribute [increases the cost of Metrics updates](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Metrics/MetricsBenchmarks.cs#L39-L45).!

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 6 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: I removed the queue and network metrics. It will only have `db.client.operation.duration`.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 8 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: makes sense, port will be added only if not the default 6379

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 9 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: @cijothomas could you resume the review to see if it meets expectations?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: sure will do today.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+using System.Reflection;
+using OpenTelemetry.Internal;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation;
+
+internal class RedisMetrics : IDisposable
+{
+    internal const string DurationMetricName = "db.client.operation.duration";
+    internal const string QueueTimeMetricName = "db.client.operation.queue_time";
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:14]: you can remove unused strings from this file.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    internal const string ServerTimeMetricName = "db.client.operation.server_time";
+
+    internal static readonly Assembly Assembly = typeof(StackExchangeRedisInstrumentation).Assembly;
+    internal static readonly AssemblyName AssemblyName = Assembly.GetName();
+    internal static readonly string InstrumentationName = AssemblyName.Name!;
+    internal static readonly string InstrumentationVersion = Assembly.GetPackageVersion();
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:20]: Any reason why is this diff. than https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs#L27 ? Not sure which one is apt here, but its best to stick with existing practices.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+    public RedisMetrics()
+    {
+        this.meter = new Meter(InstrumentationName, InstrumentationVersion);
+
+        this.DurationHistogram = this.meter.CreateHistogram<double>(
+            DurationMetricName,
+            unit: "s",
+            description: "Total client request duration, including processing, queue and server duration.",
+            [new(SemanticConventions.AttributeDbSystem, "redis")]);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:32]: I love this optimization of putting this tag in the Meter level and it definitely helps save perf. I am not 100% sure if semantic convention allows/prohibits it. I'll check and update!

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 13]
```diff
@@ -88,30 +95,37 @@ static bool GetCommandAndKey(
...
-        if (activity == null)
+        if (activity is null && metrics.Enabled is false)
         {
             return null;
         }
 
-        activity.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        activity?.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        var meterTags = metrics.Enabled ?
+            new TagList(StackExchangeRedisConnectionInstrumentation.CreationTags) :
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:105]: creationtags is no longer needed, as it is already covered via Meter Tags.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 14]
```diff
@@ -88,30 +95,37 @@ static bool GetCommandAndKey(
...
-            // command.CreationToEnqueued;      // 00:00:32.4571995
-            // command.EnqueuedToSending;       // 00:00:00.0352838
-            // command.SentToResponse;          // 00:00:00.0060586
-            // command.ResponseToCompletion;    // 00:00:00.0002601
+        // Total:
+        // command.ElapsedTime;             // 00:00:32.4988020
 
-            // Total:
-            // command.ElapsedTime;             // 00:00:32.4988020
+        var flags = command.Flags.ToString();
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:121]: nit: you can revert this line, to make it easier to review the actual changes!

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 15]
```diff
@@ -88,30 +95,37 @@ static bool GetCommandAndKey(
...
+        activity?.SetTag(SemanticConventions.AttributeDbRedisFlagsKeyName, flags);
 
-            activity.SetTag(StackExchangeRedisConnectionInstrumentation.RedisFlagsKeyName, command.Flags.ToString());
+        var operationName = command.Command ?? string.Empty;
+        meterTags?.Add(SemanticConventions.AttributeDbOperationName, operationName);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:125]: `AttributeDbOperationName` was never set for Activity though that is required there. Need to check why was that the case before adding it to Metrics.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 16]
```diff
@@ -88,30 +95,37 @@ static bool GetCommandAndKey(
...
 
-        if (activity == null)
+        if (activity is null && metrics.Enabled is false)
         {
             return null;
         }
 
-        activity.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        activity?.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        var meterTags = metrics.Enabled ?
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:104]: `TagList tags = default;`

Isn't this sufficient?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md [hunk 17]
```diff
@@ -66,6 +68,11 @@ public class Program
             .AddRedisInstrumentation(connection)
             .AddConsoleExporter()
             .Build();
+
+        using var tracerProvider = Sdk.CreateMeterProviderBuilder()
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md:72]:
```suggestion
        using var meterProvider = Sdk.CreateMeterProviderBuilder()
```

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md [hunk 18]
```diff
@@ -66,6 +68,11 @@ public class Program
...
+            .AddRedisInstrumentation()
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md:73]:
```suggestion
            .AddRedisInstrumentation(connection)
```

**Review by cijothomas — COMMENTED** [review]
Did a general review with my limited Redis experience!
The existing trace one is probably out of date with recent conventions, it'd be good to get that fixed (separate PR) as well!

Need to get review from people with more Redis experience to take this in. @matt-hensley Could you help?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 11 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:17]: I opted for the one that I saw being used most recently. I'm also not sure which one should be used and why there are variations.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 12 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: removed, can you check, please?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 13 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:105]: since you suggest removing the meter tags https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1982#discussion_r1725675861, I will keep it

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 14 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: reverted

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 15 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs]: Here we are creating the metrics following the most recent semantic conventions, on the other hand, tracing still follows an old version. I think it's better to keep the activities as is and handle it in a separate PR, as you suggested here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1982#pullrequestreview -2251326101

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 11 above)
Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:17]: `Assemlby.GetPackageVersion()` is correct. `AssemblyName.Version!.ToString();` always return 1.0.0.0.
As I remember, we have decided to keep it as is, due to migration to natively implemented metrics by .NET team.

Ref: #1624

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md (hunk 18 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md]: Done

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md (hunk 17 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md]: Done

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs (hunk 10 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs]: Done

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs (hunk 16 above)
Comment (tiagodaraujo) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:104]: I think the result of `default(TagList)` is an empty TagList, am I wrong?

Here I would like to set a null object to be validated in every AddTag and so optimize the memory allocation

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/StackExchangeRedisInstrumentation.cs [hunk 19]
```diff
@@ -20,7 +20,7 @@ internal StackExchangeRedisInstrumentation(
         this.options = options;
     }
 
-    internal List<StackExchangeRedisConnectionInstrumentation> InstrumentedConnections { get; } = new();
+    internal List<StackExchangeRedisConnectionInstrumentation> InstrumentedConnections { get; } = [];
```
Comment (ysolomchenko) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/StackExchangeRedisInstrumentation.cs:23]: Move square brackets to a new line or revert to new();

File: test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/Implementation/RedisProfilerEntryInstrumenterConfigurationTests.cs [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+namespace OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.Implementation;
+
+[Collection("Redis")]
+public class RedisProfilerEntryInstrumenterConfigurationTests
+{
+    private const int MaxTimeToAllowForFlush = 20000;
+
+    private readonly ConnectionMultiplexer connection;
+    private readonly RedisMetrics metrics;
+    private readonly List<Metric> exportedItems = [];
```
Comment (ysolomchenko) [inline test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/Implementation/RedisProfilerEntryInstrumenterConfigurationTests.cs:20]: Also applies here.

**Review by ysolomchenko — COMMENTED** [review]
Make this change to see the test results

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+    private readonly Meter meter;
+
+    public RedisMetrics()
+    {
+        this.meter = new Meter(InstrumentationName, InstrumentationVersion);
+
+        this.DurationHistogram = this.meter.CreateHistogram<double>(
+            DurationMetricName,
+            unit: "s",
+            description: "Total client request duration, including processing, queue and server duration.");
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisMetrics.cs:28]:
```suggestion
            description: "Duration of database client operations.");
```

ref: https://github.com/open-telemetry/semantic-conventions/blob/v1.27.0/docs/database/database-metrics.md?plain=1#L82

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 22]
```diff
@@ -135,31 +150,59 @@ static bool GetCommandAndKey(
...
                 }
-                else
+            }
+            else if (command.EndPoint is DnsEndPoint dnsEndPoint)
+            {
+                var host = dnsEndPoint.Host;
+                var port = dnsEndPoint.Port;
+
+                activity?.SetTag(SemanticConventions.AttributeNetPeerName, host);
+                activity?.SetTag(SemanticConventions.AttributeNetPeerPort, port);
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:179]: Candidate for separate PR - update semantic convention to 1.27.0. At least for network tags.
Maybe it will be worth to merge it before this PR? Then the final user experience and process review could be better.

Note. `net.peer.name` is deprecated. It it not only problem here.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs [hunk 23]
```diff
@@ -88,30 +95,38 @@ static bool GetCommandAndKey(
...
             return null;
         }
 
-        activity.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        activity?.SetEndTime(command.CommandCreated + command.ElapsedTime);
+        var meterTags = metrics.Enabled ?
+            new TagList(StackExchangeRedisConnectionInstrumentation.CreationTags) :
+            default(IList<KeyValuePair<string, object?>>);
 
-        if (activity.IsAllDataRequested)
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryInstrumenter.cs:98]: I think it is worth to keep this functionality. Checks should be cheaper than adding couple of tags.

**Review by Kielek — COMMENTED** [review]
First round. I need to think about public contract.

**Comment by POnakS** [comment]
@tiagodaraujo do you plan to proceed with this MR? This is great work! Can I help?

**Comment by tiagodaraujo** [comment]
@POnakS unfortunately I don't have a plan to proceed. I've had a lot of difficulty moving forward with this. I hope someone will catch on to this one day. You can copy and paste, as long as it helps the community for the common interest of all.
